src/mscorlib/src/System/Lazy.cs

```diff
@@ -202,27 +308,123 @@ public Lazy(Func<T> valueFactory, bool isThreadSafe)
         /// a null reference (Nothing in Visual Basic).</exception>
         /// <exception cref="System.ArgumentOutOfRangeException"><paramref name="mode"/> mode contains an invalid value.</exception>
         public Lazy(Func<T> valueFactory, LazyThreadSafetyMode mode)
+            : this(valueFactory, mode, false)
         {
-            if (valueFactory == null)
+        }
+
+        private Lazy(Func<T> valueFactory, LazyThreadSafetyMode mode, bool useDefaultConstructor)
+        {
+            if (valueFactory == null && !useDefaultConstructor)
                 throw new ArgumentNullException(nameof(valueFactory));
 
-            m_threadSafeObj = GetObjectFromMode(mode);
-            m_valueFactory = valueFactory;
+            _factory = valueFactory;
+            _state = LazyNonGeneric.Create(mode, useDefaultConstructor);
         }
 
-        /// <summary>
-        /// Static helper function that returns an object based on the given mode. it also throws an exception if the mode is invalid
-        /// </summary>
-        private static object GetObjectFromMode(LazyThreadSafetyMode mode)
+        private T ViaConstructor()
         {
-            if (mode == LazyThreadSafetyMode.ExecutionAndPublication)
-                return new object();
-            else if (mode == LazyThreadSafetyMode.PublicationOnly)
-                return LazyHelpers.PUBLICATION_ONLY_SENTINEL;
-            else if (mode != LazyThreadSafetyMode.None)
-                throw new ArgumentOutOfRangeException(nameof(mode), Environment.GetResourceString("Lazy_ctor_ModeInvalid"));
-            
-            return null; // None mode
+            _value = CreateViaDefaultConstructor();
+            _state = null;
+
+            return _value;
+        }
+
+        private T ViaFactory(LazyThreadSafetyMode mode)
+        {
+            try
+            {
+                var factory = _factory;
```

Comment:
Nit: please use the type name instead of `var` when the type isn't obvious from the RHS of the expression